bootstrap: unify tenant and sys tables

dt opened the pull request:
We want to be able to replicate a system tenant into a non-system tenant
but this requires that they have the same set of system tables so that
when the non-system tenant starts executing, it finds all the same tables
that it expects to exist, regardless of whether it was created via PCR of
a system tenant or bootstrap of a secondary tenant.


Release note: none.
Epic: none.

dt committed:
bootstrap: unify tenant and sys tables

We want to be able to replicate a system tenant into a non-system tenant
but this requires that they have the same set of system tables so that
when the non-system tenant starts executing, it finds all the same tables
that it expects to exist, regardless of whether it was created via PCR of
a system tenant or bootstrap of a secondary tenant.

Release note: none.
Epic: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/122005)
<!-- Reviewable:end -->

stevendanna commented on pkg/sql/catalog/bootstrap/metadata.go:
```diff
@@ -101,18 +101,12 @@ func (ms *MetadataSchema) AddDescriptor(desc catalog.Descriptor) {
-		return
-	}
 	ms.AddDescriptor(desc)
 }
```
We may want to remove these functions and use AddDescriptor directly. With this change the comments are lying a bit.

stevendanna reviewed (commented):
Some commentary in the commit message about why we think this is reasonable might be useful. I imagine there is a long-list of changes needed to make the tests pass so I'll give this another look when those have settled a bit.

dt commented:
oh yeah, this was just to let CI tell me what is broken. I plan to add the migration to this (and commentary) before it is ready to go.

dt committed:
upgrades: add the span_count table to sys tenants

They need it if they will be replicated via PCR later so we create it
during the upgrade if it is missing. It was previously added to bootstrap.

Epic: none.
Release note: none.

dt commented:
> Some commentary in the commit message about why we think this is reasonable might be useful.

Done.

> a long-list of changes needed to make the tests pass

Done.

dt commented:
TFTR

bors r+

dt commented:
bors r+

dt commented:
blathers backport 24.1
